In [1]:
import pandas as pd
import numpy as np

# Eurostat datasets inspection

This notebook provides a sequential inspection of datasets selected from Eurostat, focusing on their structure, data quality, temporal and geographic granularity, value ranges, units, and consistency across datasets.

## **nrg_ind_ren**: Cuota de energía renovable sobre el consumo bruto final por país

In [2]:
ind_ren_df = pd.read_csv(r"../data_raw/nrg_ind_ren.csv", sep=',', encoding='utf-8')
ind_ren_df.head(5)

,freq,nrg_bal,unit,geo\TIME_PERIOD,2004,2005,2006,2007,2008,2009,...,2016,2017,2018,2019,2020,2021,2022,2023,2024,2025
0,A,REN,PC,AL,29.620,31.367,32.070,32.657,32.448,31.437,...,36.953,35.776,36.572,38.042,45.015,41.389,44.076,46.620,43.111,NaN
1,A,REN,PC,AT,22.553,24.353,26.276,28.144,28.788,31.039,...,33.370,33.136,33.784,33.755,36.545,34.643,34.057,41.600,42.951,44.2
2,A,REN,PC,BA,NaN,NaN,NaN,NaN,NaN,NaN,...,25.358,23.241,35.972,37.454,39.835,36.587,39.320,40.038,38.524,NaN
3,A,REN,PC,BE,1.916,2.325,2.658,3.140,3.611,4.746,...,8.744,9.136,9.472,9.929,13.000,13.017,13.831,14.673,14.336,14.9
4,A,REN,PC,BG,9.231,9.173,9.415,9.098,10.345,12.005,...,18.760,18.695,20.581,21.545,23.319,19.394,19.079,23.151,23.210,22.7


In [3]:
ind_ren_df.describe()

,2004,2005,2006,2007,2008,2009,2010,2011,2012,2013,...,2016,2017,2018,2019,2020,2021,2022,2023,2024,2025
count,140.000000,144.000000,144.000000,144.000000,144.000000,144.000000,144.000000,144.000000,144.000000,144.000000,...,148.000000,148.000000,148.000000,148.000000,148.000000,167.00000,167.000000,167.000000,167.000000,63.000000
mean,14.551821,15.530965,15.766257,16.533785,17.032667,18.732458,19.158403,19.694201,20.926778,21.761743,...,23.726480,24.177770,25.204162,26.190007,28.150764,28.18718,29.701970,31.558695,32.201126,40.036508
std,18.140058,18.616635,18.596241,19.538299,18.667909,19.500543,19.343394,19.895495,20.118372,20.014927,...,20.767711,20.801582,21.185184,21.443619,21.945317,21.85074,23.069077,23.577290,23.686296,21.970584
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,11.200000
25%,1.560750,1.800750,2.100000,2.907250,3.592250,4.566500,5.169500,4.716250,5.501250,6.176500,...,7.193500,7.448000,8.302000,8.913250,11.667000,12.04200,12.090000,13.944000,14.332500,23.000000
50%,7.646000,8.555000,9.046000,9.923000,10.834000,12.495000,12.882500,13.639500,15.743000,16.700000,...,18.402000,19.271000,20.319000,21.756500,23.192500,22.00100,22.901000,25.066000,25.378000,34.300000
75%,20.249000,23.035500,23.737250,24.859250,25.877500,28.253250,29.589750,31.066500,31.911250,32.688250,...,32.550250,34.174750,35.178250,36.542750,37.576250,38.79450,41.303500,43.489000,44.323000,53.850000
max,97.973000,97.423000,100.843000,113.700000,100.154000,105.191000,98.252000,105.887000,104.599000,106.917000,...,105.638000,104.850000,106.829000,110.445000,113.802000,113.64200,120.010000,116.538000,115.063000,122.000000


In [4]:
ind_ren_df.shape

(167, 26)

In [5]:
ind_ren_df.dtypes

freq                object
nrg_bal             object
unit                object
geo\TIME_PERIOD     object
2004               float64
2005               float64
2006               float64
2007               float64
2008               float64
2009               float64
2010               float64
2011               float64
2012               float64
2013               float64
2014               float64
2015               float64
2016               float64
2017               float64
2018               float64
2019               float64
2020               float64
2021               float64
2022               float64
2023               float64
2024               float64
2025               float64
dtype: object

In [6]:
ind_ren_df.isna().sum()

freq                 0
nrg_bal              0
unit                 0
geo\TIME_PERIOD      0
2004                27
2005                23
2006                23
2007                23
2008                23
2009                23
2010                23
2011                23
2012                23
2013                23
2014                19
2015                19
2016                19
2017                19
2018                19
2019                19
2020                19
2021                 0
2022                 0
2023                 0
2024                 0
2025               104
dtype: int64

In [7]:
ind_ren_df.duplicated().sum()

np.int64(0)

In [8]:
ind_ren_df['geo\\TIME_PERIOD'].duplicated().sum()

np.int64(129)

In [9]:
mask = ind_ren_df.duplicated(
    subset=['geo\\TIME_PERIOD'],
    keep=False
)

ind_ren_df[mask].sort_values(
    ['geo\\TIME_PERIOD']
).head(10)

,freq,nrg_bal,unit,geo\TIME_PERIOD,2004,2005,2006,2007,2008,2009,...,2016,2017,2018,2019,2020,2021,2022,2023,2024,2025
0,A,REN,PC,AL,29.620,31.367,32.070,32.657,32.448,31.437,...,36.953,35.776,36.572,38.042,45.015,41.389,44.076,46.620,43.111,NaN
76,A,REN_HEAT_CL,PC,AL,33.138,37.757,30.964,33.150,37.056,34.730,...,32.495,23.969,21.993,24.754,25.125,21.200,22.087,22.917,3.368,NaN
121,A,REN_TRA,PC,AL,0.099,0.095,0.000,0.000,0.000,0.000,...,0.000,0.000,0.000,0.203,0.347,0.344,0.402,0.456,0.217,NaN
38,A,REN_ELC,PC,AL,69.979,76.112,74.208,79.631,73.299,70.714,...,82.104,91.025,92.450,92.994,99.975,94.421,102.969,105.381,105.188,NaN
1,A,REN,PC,AT,22.553,24.353,26.276,28.144,28.788,31.039,...,33.370,33.136,33.784,33.755,36.545,34.643,34.057,41.600,42.951,44.2
77,A,REN_HEAT_CL,PC,AT,20.165,22.817,24.491,27.142,27.195,29.626,...,33.480,33.670,34.178,33.929,34.995,33.127,31.819,40.770,42.169,NaN
39,A,REN_ELC,PC,AT,61.627,62.901,63.527,65.666,65.880,68.626,...,72.521,71.626,74.189,75.067,78.204,74.008,73.640,87.899,90.065,90.8
122,A,REN_TRA,PC,AT,4.532,5.070,7.532,8.241,9.560,11.175,...,10.583,9.703,9.931,10.051,10.283,9.648,10.751,13.576,14.591,NaN
40,A,REN_ELC,PC,BA,NaN,NaN,NaN,NaN,NaN,NaN,...,40.829,39.564,41.498,45.482,49.308,46.420,48.521,51.667,52.742,NaN
78,A,REN_HEAT_CL,PC,BA,NaN,NaN,NaN,NaN,NaN,NaN,...,32.329,28.739,52.685,55.873,56.623,53.310,57.648,59.672,56.596,NaN


In [10]:
ind_ren_df['nrg_bal'].unique()

array(['REN', 'REN_ELC', 'REN_HEAT_CL', 'REN_HEAT_CL_WHC', 'REN_TRA',
       'REN_WHC_DHEAT_DCL'], dtype=object)

In [11]:
ind_ren_df[ind_ren_df['geo\\TIME_PERIOD'] == 'PT']

,freq,nrg_bal,unit,geo\TIME_PERIOD,2004,2005,2006,2007,2008,2009,...,2016,2017,2018,2019,2020,2021,2022,2023,2024,2025
31,A,REN,PC,PT,19.205,19.523,20.792,21.907,22.929,24.405,...,30.864,30.611,30.203,30.623,33.982,33.982,34.675,35.154,36.319,36.7
69,A,REN_ELC,PC,PT,27.390,27.703,29.308,32.285,34.063,37.562,...,53.990,54.168,52.186,53.774,58.032,58.433,60.991,63.010,65.801,65.6
107,A,REN_HEAT_CL,PC,PT,32.490,32.075,34.234,34.951,37.451,37.931,...,41.614,41.022,40.924,41.657,41.546,42.681,45.545,47.098,46.952,NaN
152,A,REN_TRA,PC,PT,0.424,0.450,1.631,2.433,2.515,3.885,...,7.647,7.914,9.037,9.089,9.701,8.611,8.706,11.162,14.329,NaN


In [12]:
ind_ren_df[['nrg_bal', 'geo\\TIME_PERIOD']].duplicated().sum()

np.int64(0)

## Observaciones para el dataset 'nrg_ind_ren'

En este primer dataset de Eurostat se muestra el porcentaje sobre el total de energía consumida en bruto que corresponde a fuentes de energía renovable, las propias unidades quedan confirmadas dentro de los propios datos por la columna 'unit'.

Se trata de un dataset corto de apenas 167 filas y 26 columnas, muchas de ellas relativas al año para el cual se muestra la métrica correspondiente.

Se observan distintos valores nulos por columnas excepto para el periodo de 2021 a 2024.
Es posible que el repunte de nulos de 2025 (104 de 167 filas) venga dado por estar aún actualizándose; habrá que decidir si obviar este año una vez se confirme la hipótesis.
Para el resto de años lo más prudente será si decidir entre eliminar aquellos países con menos influencia en el análisis que presenten nulos o si simplemente dejarlos como valores nulos.

Si bien no hay duplicados globales, vemos que la granularidad no es simplemente por país, de modo que el dataset no muestra un único valor porcentual de consumo de energía renovable sobre consumo total.
Se comprueba que para cada país hay un desglose de hasta un máximo de 6 subcategorías en función del balance/sector (nrg_bal).
Para cada país se presenta el valor porcentual de consumo renovable sobre consumo bruto total con el código 'REN', luego se subdivide este consumo en otros porcentajes que lo conforman:

'REN_ELC': Muestra el porcentaje de renovables sobre el consumo total solo de electricidad.
'REN_HEAT_CL': En este caso muestra el porcentaje sobre el total de calefacción y refrigeración.
'REN_TRA': Aplica sobre el consumo total en transporte.
'REN_HEAT_CL_WHC' y 'REN_WHC_DHEAT_DCL' están relacionados con el consumo de frío/calor residual.

Para este proyecto es muy probable que simplemente se use el valor genérico de 'REN' por fecha y país, dicho lo cual, hasta que se cree un segundo dataset que solo tenga en cuenta estos factores, la granularidad del dataset se mantiene en país y balance/sector del consumo, siendo el objetivo que la granularidad acabe marcada por el año (ahora son columnas), país y balance/sector.

Cabe la posibilidad de usar 'REN_ELC' como variable adicional dentro de la hipótesis número 5 dado su buen encaje con el dataset '26-centrais' de E-Redes.

El tipo de datos en este caso debería ser distinto para 'nrg_bal' y 'geo\TIME_PERIOD', las cuales deberían ser categóricas, aunque la segunda podría ser texto. Por otro lado, unit puede ser categórica ya que al ser 'PC' para todos los registros, no aporta variabilidad.

La franja temporal de este dataset abarca desde 2004 a 2025, a la espera de decidir si el corte aplica a finales de 2024 por falta de datos en 2025.

## **nrg_pc_204**: Precios de la electricidad para consumo desde 2007

In [13]:
nrg_pc_df = pd.read_csv(r"../data_raw/nrg_pc_204.csv", sep=',', encoding='utf-8')
nrg_pc_df.head(5)

,freq,siec,nrg_cons,unit,tax,currency,geo\TIME_PERIOD,2007-S1,2007-S2,2008-S1,...,2021-S1,2021-S2,2022-S1,2022-S2,2023-S1,2023-S2,2024-S1,2024-S2,2025-S1,2025-S2
0,S,E7000,KWH1000-2499,KWH,I_TAX,EUR,AL,NaN,NaN,NaN,...,0.0925,0.0937,0.0941,0.0976,0.1010,0.1089,0.1114,0.1150,0.1156,0.1175
1,S,E7000,KWH1000-2499,KWH,I_TAX,EUR,AT,NaN,0.2008,0.2005,...,0.2650,0.2787,0.2616,0.2614,0.2988,0.3055,0.3011,0.2631,0.3165,0.3669
2,S,E7000,KWH1000-2499,KWH,I_TAX,EUR,BA,NaN,NaN,NaN,...,0.0980,0.0970,0.0996,0.1003,0.1043,0.0990,0.1067,0.1054,0.1100,0.1125
3,S,E7000,KWH1000-2499,KWH,I_TAX,EUR,BE,NaN,0.1867,0.2172,...,0.2910,0.3161,0.3633,0.4586,0.4559,0.4111,0.3715,0.3695,0.3939,0.3806
4,S,E7000,KWH1000-2499,KWH,I_TAX,EUR,BG,NaN,0.0721,0.0731,...,0.1045,0.1104,0.1110,0.1147,0.1146,0.1192,0.1193,0.1224,0.1313,0.1355


In [14]:
nrg_pc_df.describe()

,2007-S1,2007-S2,2008-S1,2008-S2,2009-S1,2009-S2,2010-S1,2010-S2,2011-S1,2011-S2,...,2021-S1,2021-S2,2022-S1,2022-S2,2023-S1,2023-S2,2024-S1,2024-S2,2025-S1,2025-S2
count,294.000000,1341.000000,1380.000000,1380.000000,1389.000000,1389.000000,1398.000000,1413.000000,1503.000000,1503.000000,...,1833.000000,2106.000000,2061.000000,2106.000000,2097.000000,2106.000000,2052.000000,2097.000000,2061.000000,2052.000000
mean,0.394403,0.604298,0.674864,0.685397,0.658901,0.687033,0.708438,0.686232,0.799585,0.799596,...,0.981787,0.966664,1.020127,1.144810,1.196881,1.177070,1.191315,1.221705,1.255416,1.056202
std,0.774067,3.136053,3.727254,3.657299,3.805763,4.104914,4.228886,4.020796,4.222051,4.203328,...,4.059104,3.906895,3.992188,4.533525,4.814490,4.802309,4.933486,4.958421,5.078791,4.271281
min,0.054400,0.018800,0.044400,0.057600,0.051700,0.047900,0.058500,0.058500,0.060300,0.052600,...,-0.240200,-0.237800,-0.918100,-3.514400,-0.224300,-0.224200,-0.029000,-0.098400,0.048300,0.046900
25%,0.118300,0.104200,0.108075,0.118375,0.120400,0.119100,0.120675,0.123200,0.121750,0.124300,...,0.121500,0.135225,0.143300,0.163725,0.162100,0.161400,0.160150,0.158300,0.162100,0.165800
50%,0.158600,0.137400,0.141250,0.156550,0.153700,0.154300,0.158450,0.161700,0.160900,0.165100,...,0.168700,0.183100,0.198500,0.236000,0.242400,0.233900,0.227750,0.226600,0.227000,0.230250
75%,0.227975,0.188600,0.197500,0.210350,0.211500,0.210700,0.218175,0.219400,0.221000,0.227550,...,0.255500,0.262475,0.300500,0.355000,0.362700,0.333100,0.316825,0.313200,0.313100,0.325200
max,5.930000,37.920000,40.990000,40.330000,44.470000,49.240000,53.610000,49.690000,50.842800,52.221700,...,48.911600,49.075700,50.349700,53.885900,62.657100,64.287600,62.100100,63.600000,66.530000,61.179200


In [15]:
nrg_pc_df['unit'].unique()

array(['KWH'], dtype=object)

In [16]:
nrg_pc_df[nrg_pc_df['geo\\TIME_PERIOD'] == 'PT'].head(18)

,freq,siec,nrg_cons,unit,tax,currency,geo\TIME_PERIOD,2007-S1,2007-S2,2008-S1,...,2021-S1,2021-S2,2022-S1,2022-S2,2023-S1,2023-S2,2024-S1,2024-S2,2025-S1,2025-S2
33,S,E7000,KWH1000-2499,KWH,I_TAX,EUR,PT,NaN,0.1552,0.1695,...,0.2294,0.2376,0.2429,0.2461,0.2264,0.2447,0.2632,0.2877,0.2707,0.2773
75,S,E7000,KWH1000-2499,KWH,I_TAX,NAC,PT,NaN,0.1552,0.1695,...,0.2294,0.2376,0.2429,0.2461,0.2264,0.2447,0.2632,0.2877,0.2707,0.2773
115,S,E7000,KWH1000-2499,KWH,I_TAX,PPS,PT,NaN,0.1908,0.2040,...,0.2696,0.2774,0.2836,0.2874,0.2648,0.2902,0.3183,0.3496,0.3283,0.3363
156,S,E7000,KWH1000-2499,KWH,X_TAX,EUR,PT,NaN,0.1180,0.1242,...,0.1267,0.1311,0.1900,0.2325,0.3059,0.2351,0.2013,0.1868,0.1818,0.1863
198,S,E7000,KWH1000-2499,KWH,X_TAX,NAC,PT,NaN,0.1180,0.1242,...,0.1267,0.1311,0.1900,0.2325,0.3059,0.2351,0.2013,0.1868,0.1818,0.1863
238,S,E7000,KWH1000-2499,KWH,X_TAX,PPS,PT,NaN,0.1451,0.1495,...,0.1489,0.1531,0.2219,0.2715,0.3578,0.2788,0.2434,0.2270,0.2204,0.2260
279,S,E7000,KWH1000-2499,KWH,X_VAT,EUR,PT,NaN,0.1478,0.1615,...,0.1943,0.2055,0.2067,0.2112,0.1968,0.2151,0.2302,0.2518,0.2410,0.2472
321,S,E7000,KWH1000-2499,KWH,X_VAT,NAC,PT,NaN,0.1478,0.1615,...,0.1943,0.2055,0.2067,0.2112,0.1968,0.2151,0.2302,0.2518,0.2410,0.2472
361,S,E7000,KWH1000-2499,KWH,X_VAT,PPS,PT,NaN,0.1817,0.1943,...,0.2284,0.2399,0.2414,0.2467,0.2302,0.2551,0.2784,0.3059,0.2922,0.2998
402,S,E7000,KWH2500-4999,KWH,I_TAX,EUR,PT,NaN,0.1562,0.1482,...,0.2089,0.2170,0.2199,0.2222,0.2071,0.2299,0.2426,0.2626,0.2390,0.2435


In [17]:
nrg_pc_df.shape

(2196, 45)

In [18]:
nrg_pc_df.dtypes

freq                object
siec                object
nrg_cons            object
unit                object
tax                 object
currency            object
geo\TIME_PERIOD     object
2007-S1            float64
2007-S2            float64
2008-S1            float64
2008-S2            float64
2009-S1            float64
2009-S2            float64
2010-S1            float64
2010-S2            float64
2011-S1            float64
2011-S2            float64
2012-S1            float64
2012-S2            float64
2013-S1            float64
2013-S2            float64
2014-S1            float64
2014-S2            float64
2015-S1            float64
2015-S2            float64
2016-S1            float64
2016-S2            float64
2017-S1            float64
2017-S2            float64
2018-S1            float64
2018-S2            float64
2019-S1            float64
2019-S2            float64
2020-S1            float64
2020-S2            float64
2021-S1            float64
2021-S2            float64
2

In [19]:
nrg_pc_df.isna().sum()

freq                  0
siec                  0
nrg_cons              0
unit                  0
tax                   0
currency              0
geo\TIME_PERIOD       0
2007-S1            1902
2007-S2             855
2008-S1             816
2008-S2             816
2009-S1             807
2009-S2             807
2010-S1             798
2010-S2             783
2011-S1             693
2011-S2             693
2012-S1             612
2012-S2             567
2013-S1             492
2013-S2             492
2014-S1             462
2014-S2             462
2015-S1             432
2015-S2             432
2016-S1             402
2016-S2             402
2017-S1             396
2017-S2             375
2018-S1             390
2018-S2             336
2019-S1             345
2019-S2             336
2020-S1             327
2020-S2             363
2021-S1             363
2021-S2              90
2022-S1             135
2022-S2              90
2023-S1              99
2023-S2              90
2024-S1         

In [20]:
nrg_pc_df.duplicated().sum()

np.int64(0)

In [21]:
nrg_pc_df[['geo\\TIME_PERIOD', 'nrg_cons', 'tax', 'currency']].duplicated().sum()

np.int64(0)

## Observaciones para el dataset 'nrg_pc_204'

Las unidades de las métricas vienen incluídas en la columna 'unit' siendo que los precios se expresan en kWh.

El dataset presenta 2196 filas y 45 columnas, muchas de ellas muestran métricas semestrales en un rango de 18 años.

Se observa que la columna 'nrg_cons' indica la banda de consumo que abarca las siguientes opciones:

KWH_LT1000

KWH1000-2499

KWH2500-4999

KWH5000-14999

KWH_GE15000

TOT_KWH

Es de mucha utilidad tener el desglose por franjas además del total puesto que se analilzará el tipo de consumo y no solo la cantidad.

La columna 'tax' indica el tipo de impuesto incluido en el precio marcado, 'I_TAX' incluye todos los impuestos y tasas, 'X_TAX' no incluye impuestos y 'X_VAT' incluye impuestos y tasas pero no el IVA.

Para este proyecto se filtrará de forma que se tenga únicamente en cuenta el precio con impuestos, dado que el objetivo es determinar el precio que paga realmente el consumidor portugués.

Por otro lado, la columna currency indica el valor en euros y en la moneda nacional (en este caso coincide). Además hay una tercera opción que es el 'PPS' que indica el 'Purchasing Power Standard', una unidad ajustada por poder adquisitivo que es muy útil para hacer comparativas entre países.

En cuanto a tipos, cabría transformar a categóricas las columnas 'freq', 'siec' 'nrg_cons' (en este caso con orden), 'unit', 'tax', 'currency' y 'geo\\TIME_PERIOD'.

'freq', 'siec' y 'unit' se mantienen constantes por lo que no son muy relevantes. En cuanto a 'nrg_cons', habría que tener en cuenta que 'TOT_KWH' es un agregado.


Se observan bastantes nulos para las columnas con métricas y ninguno para las demás. Si bien se puede valorar reducir el rango temporal en función de como avance el análisis, la idea sería mantener los nulos o estudiar si hay alguna forma de imputarlos que sea fiable.

No hay duplicados globales y se determina que las dimensiones que marcan la granularidad son 'geo\\TIME_PERIOD', 'ngr_cons', 'tax' y 'currency'. Si al final se pasara a formato largo la granularidad sería de 'periodo', 'país', 'nrg_cons', 'tax' y 'currency', aunque no será el objetivo inicialmente.
